Document how users can use linting tools instead of `kedro lint` (#1904)
kedro-org/kedro: docs/source/development/linting.md

# Linting

## Introduction

> Lint, or a linter, is a static code analysis tool used to flag programming errors, bugs, stylistic errors and suspicious constructs.

Reviewer: This line refers specifically to the software 'Lint' (https://en.wikipedia.org/wiki/Lint_(software)), so I don't think it is appropriate to use this here.

I also don't think that 'suspicious constructs' makes too much sense in the context of Python.

```suggestion
Linting tools are used to improve the quality of your code by checking for errors and making sure it meets a stylistic standard before it is run.
```

As a project grows and goes through various stages of development, it becomes important to maintain the code quality. Linting makes
your code more readable, easy to debug and maintain, and stylistically consistent.
## Set up linting tools
There are a variety of linting tools available to use
with your Kedro projects. This guide shows you how to use [`black`](https://github.com/psf/black), [`flake8`](https://gitlab.com/pycqa/flake8), and [`isort`](https://github.com/PyCQA/isort) to lint your Kedro projects.

> - **`black`** is a [PEP 8](https://peps.python.org/pep-0008/) compliant opinionated Python code formatter. Read the full documentation [here](https://black.readthedocs.io/en/stable/).

Author: Thank you @MerelTheisenQB for the review. I've made the suggested changes in 6c2271dc0c6ad7582a0e62950354989059690b1d.

> - **`flake8`** is a wrapper around [`pep8`](https://pypi.org/project/pep8/), [`pyflakes`](https://pypi.org/project/pyflakes/), and [`mccabe`](https://pypi.org/project/mccabe/). Read the full documentation [here](https://flake8.pycqa.org/en/latest/).

Reviewer 2: ```suggestion
- **`flake8`** is a wrapper around [`pep8`](https://pypi.org/project/pep8/), [`pyflakes`](https://pypi.org/project/pyflakes/), and [`mccabe`](https://pypi.org/project/mccabe/). [You can read more in the flake8 documentation](https://flake8.pycqa.org/en/latest/).
```

> - **`isort`** is a Python library used to sort imports alphabetically. Read the full documentation [here](https://pycqa.github.io/isort/).

Reviewer: I think the introduction to `black`, `isort` and `flake8` could be a bit longer, maybe with an example each of what they can do for the reader.

### Install linting tools
You can install `black`, `flake8`, and `isort` by adding the following lines to your project's `src/requirements.txt` file:
```text
black==22.1.0 # Used for formatting code
flake8>=3.7.9, <5.0 # Used for linting code

> isort~=5.0 # Used for linting code

Author: Thank you for the review @jmholzer! I've made the suggested changes in 6c2271dc0c6ad7582a0e62950354989059690b1d

```
To install all the project-specific dependencies, including the linting tools, navigate to the root directory of the project and run:
```bash
pip install -r src/requirements.txt
```
Alternatively, you can individually install the linting tools using the following shell commands:
```bash
pip install black
pip install flake8
pip install isort
```

### Run linting tools
Use the following commands to run lint checks:
```bash
black --check <project_root>
flake8 <project_root>
isort --check <project_root>
```
You can also have `black` and `isort` automatically format your code by omitting the `--check` flag.

## Automating linting with `pre-commit` hooks

You can also automate linting by using [`pre-commit`](https://github.com/pre-commit/pre-commit) hooks.
These hooks are run before committing your code to your repositories to automatically point out formatting issues,
making code reviews easier and less time-consuming.

### Install `pre-commit`

> You can install `pre-commit` along with other dependencies by including it in the `src\requirements.txt` file of your

Reviewer 2: ```suggestion
You can install `pre-commit` along with other dependencies by including it in the `src/requirements.txt` file of your
```

Kedro project by adding the following line:
```text
pre-commit
```

> You can also install is using the following command:

Reviewer 2: ```suggestion
You can also install `pre-commit` using the following command:
```

```bash
pip install pre-commit
```

> ### Add pre-commit configuration file

Reviewer 2: ```suggestion
### Add `pre-commit` configuration file
```

maintain the code quality. Linting makes your code more readable, easy to debug and maintain, and stylistically
consistent.

## Set up linting tools
There are a variety of linting tools available to use with your Kedro projects. This guide shows you how to use
[`black`](https://github.com/psf/black), [`flake8`](https://gitlab.com/pycqa/flake8), and
[`isort`](https://github.com/PyCQA/isort) to lint your Kedro projects.
- **`black`** is a [PEP 8](https://peps.python.org/pep-0008/) compliant opinionated Python code formatter. `black` can
check for styling inconsistencies and reformat your files in place.
[You can read more in the `black` documentation](https://black.readthedocs.io/en/stable/).
- **`flake8`** is a wrapper around [`pep8`](https://pypi.org/project/pep8/),
[`pyflakes`](https://pypi.org/project/pyflakes/), and [`mccabe`](https://pypi.org/project/mccabe/) which can flag
programming errors and coding style inconsistencies with respect to [PEP 8](https://peps.python.org/pep-0008/),
and check the cyclomatic complexity of your code base.
[You can read more in the `flake8` documentation](https://flake8.pycqa.org/en/latest/).

> - **`isort`** is a Python library used to sort imports alphabetically, and automatically separated into sections and by

Reviewer 2: ```suggestion
- **`isort`** is a Python library used to sort imports alphabetically and automatically separate them into sections and by
```

Author: Made the change in 969c6dc7102657e8a145a83c2de02ae6f8004478! Thanks!

type. [You can read more in the `isort` documentation](https://pycqa.github.io/isort/).

### Install linting tools
You can install `black`, `flake8`, and `isort` by adding the following lines to your project's `src/requirements.txt`
file:
```text
black # Used for formatting code
flake8 # Used for linting code
isort # Used for linting code
```
To install all the project-specific dependencies, including the linting tools, navigate to the root directory of the
project and run:
```bash
pip install -r src/requirements.txt
```
Alternatively, you can individually install the linting tools using the following shell commands:
```bash
pip install black
pip install flake8
pip install isort
```

### Run linting tools
Use the following commands to run lint checks:
```bash
black --check <project_root>
flake8 <project_root>
isort --profile black --check <project_root>
```
You can also have `black` and `isort` automatically format your code by omitting the `--check` flag. Since `isort` and
`black` both format your imports, adding `--profile black` to the `isort` run helps avoid potential conflicts.

## Automating linting with `pre-commit` hooks

> You can also automate linting by using [`pre-commit`](https://github.com/pre-commit/pre-commit) hooks.

Reviewer 3: ```suggestion
[`pre-commit`](https://github.com/pre-commit/pre-commit) hooks can be used to automate linting.
```

Sorry, this might be a bit pedantic I think we should avoid starting with 'You can also'

Author: Thanks @SajidAlamQB for the review! 
Do you think I could remove the 'also' instead - "You can automate linting by using `pre-commit` hooks"? 
The [contribute to kedro docs](https://kedro.readthedocs.io/en/stable/contribution/documentation_contributor_guidelines.html#style) guidelines mention avoiding passive tense.

Reviewer 3: Yes, that seems good as well.

Author: Updated in 969c6dc7102657e8a145a83c2de02ae6f8004478!
